File: iso27001-nestjs/src/infrastructure/telemetry/error-budget.tracker.ts
/**
 * A.17: Error Budget Tracker — SLA enforcement.
 *
 * Tracks availability against a configured error budget. When the budget is
 * exhausted the system should freeze risky deployments and trigger alerts.
 *
 * SLA mapping:
 *   99.9%  → 43.8 min downtime budget per month
 *   99.95% → 21.9 min downtime budget per month
 *   99.99% →  4.4 min downtime budget per month
 */
export interface ErrorBudgetSnapshot {
  slaTarget: number;
  totalRequests: number;
  failedRequests: number;
  clientErrors: number;
  observedAvailability: number;
  budgetConsumedPct: number;
  budgetExhausted: boolean;
}

export class ErrorBudgetTracker {
  private total = 0;
  private failed = 0;   // 5xx — consumes budget
  private clientErr = 0; // 4xx — tracked separately

  constructor(private readonly slaTarget: number = 0.999) {
    if (slaTarget <= 0 || slaTarget >= 1) {
      throw new Error('slaTarget must be between 0 and 1 exclusive');
    }
  }

  record(statusCode: number): void {
    this.total += 1;
    if (statusCode >= 500) {
      this.failed += 1;
    } else if (statusCode >= 400) {
      this.clientErr += 1;
    }
  }

  snapshot(): ErrorBudgetSnapshot {
    if (this.total === 0) {
      return {
        slaTarget: this.slaTarget,
        totalRequests: 0,
        failedRequests: 0,
        clientErrors: 0,
        observedAvailability: 1.0,
        budgetConsumedPct: 0.0,
        budgetExhausted: false,
      };
    }
    const availability = (this.total - this.failed) / this.total;
    const allowedErrorRate = 1.0 - this.slaTarget;
    const actualErrorRate = this.failed / this.total;
    const budgetConsumedPct = allowedErrorRate === 0
      ? (this.failed > 0 ? 100.0 : 0.0)
      : Math.min((actualErrorRate / allowedErrorRate) * 100.0, 100.0);

    const rounded = Math.round(budgetConsumedPct * 100) / 100;
    return {
      slaTarget: this.slaTarget,
      totalRequests: this.total,
      failedRequests: this.failed,
      clientErrors: this.clientErr,
      observedAvailability: Math.round(availability * 1_000_000) / 1_000_000,
      budgetConsumedPct: rounded,
      budgetExhausted: rounded >= 100.0,
    };
  }

  reset(): void {
    this.total = 0;
    this.failed = 0;
    this.clientErr = 0;
  }
}

/** Module-level singleton — shared across middleware */
export const errorBudget = new ErrorBudgetTracker(0.999);

File: iso27001-springboot/src/main/java/com/iso27001/api/infrastructure/telemetry/ErrorBudgetTracker.java
package com.iso27001.api.infrastructure.telemetry;

import org.springframework.stereotype.Component;
import java.util.concurrent.atomic.AtomicLong;

/**
 * A.17 — Error budget tracker (99.9% SLA).
 * 5xx responses consume budget; 4xx tracked separately but do not count against SLA.
 */
@Component
public class ErrorBudgetTracker {

    private static final double SLA_TARGET = 0.999;

    private final AtomicLong total = new AtomicLong(0);
    private final AtomicLong failed = new AtomicLong(0);   // 5xx
    private final AtomicLong clientErrors = new AtomicLong(0); // 4xx

    public void record(int statusCode) {
        total.incrementAndGet();
        if (statusCode >= 500) {
            failed.incrementAndGet();
        } else if (statusCode >= 400) {
            clientErrors.incrementAndGet();
        }
    }

    public record Snapshot(
        double slaTarget,
        long totalRequests,
        long failedRequests,
        long clientErrors,
        double observedAvailability,
        double budgetConsumedPct,
        boolean budgetExhausted
    ) {}

    public Snapshot snapshot() {
        long t = total.get();
        long f = failed.get();
        long ce = clientErrors.get();

        double availability = t == 0 ? 1.0 : (double)(t - f) / t;
        // Compute consumed budget in integer-scaled arithmetic first to avoid
        // IEEE 754 precision loss (e.g. 1/1000 / 0.001 * 100 ≈ 99.9999…).
        // consumed% = (f / t) / (1 - SLA) * 100  =  f * 100_000 / (t * (1 - SLA) * 1000)
        // Simplified: consumed = f * 100.0 / (t * allowedErrorRate)
        double allowedErrorRate = 1.0 - SLA_TARGET;        // 0.001 for 99.9 % SLA
        double consumed = (t == 0 || allowedErrorRate == 0)
            ? 0.0
            : Math.round((double) f * 100_000.0 / (t * allowedErrorRate)) / 1000.0;

        return new Snapshot(SLA_TARGET, t, f, ce, availability, consumed, consumed >= 100.0);
    }

    public void reset() {
        total.set(0);
        failed.set(0);
        clientErrors.set(0);
    }
}
